File: examples/py/example2.py
#!/usr/bin/python

import pydsrc
import sys


def compress_file(infilename_, outfilename_):
    # open existing FASTQ file
    fqfile = pydsrc.FastqFileRecordsReader()
    fqfile.Open(infilename_)

    # define compression configuration settings for DSRC module:
    # lossless mode (by default) with 'fast' performance setting
    # (corresponds to '-m0' switch in DSRC binary)
    settings = pydsrc.CompressionSettings()
    settings.DnaCompressionLevel = 0
    settings.QualityCompressionLevel = 0
    settings.FastqBufferSizeMb = 8

    # create DSRC archive
    archive = pydsrc.DsrcArchiveRecordsWriter()
    archive.StartCompress(outfilename_, settings)

    # read records from FASTQ file writing to DSRC archive
    rc = 0
    rec = pydsrc.FastqRecord()
    while fqfile.ReadNextRecord(rec):
        archive.WriteNextRecord(rec)
        rc += 1

    archive.FinishCompress()

    fqfile.Close()

    print "Success!\nRecords written: %d" % rc


def compress_file_lossy(infilename_, outfilename_):
    # open existing FASTQ file
    fqfile = pydsrc.FastqFileRecordsReader()
    fqfile.Open(infilename_)

    # define compression configuration settings for DSRC module:
    # use 'max' ratio compression mode
    # (corresponds to '-m2' switch in DSRC binary)
    settings = pydsrc.CompressionSettings()
    settings.DnaCompressionLevel = 3
    settings.QualityCompressionLevel = 2
    settings.FastqBufferSizeMb = 256

    # set lossy mode with Illumina 8-binning qualities scheme
    settings.LossyQualityCompression = True

    # keep only 2 first tokens in read tag field
    settings.TagFieldFilterMask = pydsrc.FieldMask().AddField(1).AddField(2).GetMask()

    # create DSRC archive
    archive = pydsrc.DsrcArchiveRecordsWriter()
    archive.StartCompress(outfilename_, settings)

    # read records from FASTQ file writing to DSRC archive
    rc = 0
    rec = pydsrc.FastqRecord()
    while fqfile.ReadNextRecord(rec):
        archive.WriteNextRecord(rec)
        rc += 1

    archive.FinishCompress()

    fqfile.Close()

    print "Success!\nRecords written: %d" % rc



def decompress_file(infilename_, outfilename_):
    # open existing DSRC archive file
    archive = pydsrc.DsrcArchiveRecordsReader()
    archive.StartDecompress(infilename_)

    # create results FASTQ file
    fqfile = pydsrc.FastqFileRecordsWriter()
    fqfile.Open(outfilename_)

    # read all records from DSRC archive writing to FASTQ file
    rc = 0
    rec = pydsrc.FastqRecord()
    while archive.ReadNextRecord(rec):
        fqfile.WriteNextRecord(rec)
        rc += 1

    archive.FinishDecompress()

    fqfile.Close()

    print "Success!\nRecords written: %d" % rc


if __name__ == "__main__":
    if not (len(sys.argv) == 4 or (len(sys.argv) == 5 and sys.argv[2] == "L")) \
       or not (sys.argv[1] == "c" or sys.argv[1] == "d"):
        print "usage: example2 <c|d> [L] <input file> <output file>"
        exit(-1)

    lossyMode = (len(sys.argv) == 5) and sys.argv[2] == "L"

    try:
        if sys.argv[1] == "c":
            if lossyMode:
                compress_file_lossy(sys.argv[-2], sys.argv[-1])
            else:
                compress_file(sys.argv[-2], sys.argv[-1])
        else:
            decompress_file(sys.argv[-2], sys.argv[-1])

    except Exception as e:
        print e.message
        exit(-1)

File: examples/cpplib/example2.cpp
#include <iostream>
#include <Dsrc.h>

using namespace dsrc::lib;


void PrintMessage();
int CompressFile(const char* inFilename_, const char* outFilename_);
int CompressFile_lossy(const char* inFilename_, const char* outFilename_);
int DecompressFile(const char* inFilename_, const char* outFilename_);


int main(int argc_, char* argv_[])
{
	if ( !(argc_ == 4 || argc_ == 5) || !(argv_[1][0] == 'c' || argv_[1][0] == 'd') )
	{
		PrintMessage();
		return -1;
	}

	char option = argv_[1][0];
	bool lossyMode = argc_ == 5 && argv_[2][0] == 'L';
	const char* inFilename = argv_[argc_-2];
	const char* outFilename = argv_[argc_-1];

	if (option == 'c')
	{
		if (lossyMode)
			return CompressFile_lossy(inFilename, outFilename);
		else
			return CompressFile(inFilename, outFilename);
	}
	else if (option == 'd')
	{
		return DecompressFile(inFilename, outFilename);
	}

	return -1;
}

void PrintMessage()
{
	std::cerr << "usage: example1 <c|d> [L] <in_filename> <out_filename>" << std::endl;
}

int CompressFile(const char* inFilename_, const char* outFilename_)
{
	// 1. Open FASTQ file
	//
	FastqFile fastqFile;
	if (!fastqFile.Open(inFilename_))		// TODO: analyze the dataset type
	{
		std::cerr << "Error: cannot open FASTQ file: " << inFilename_ << std::endl;
		return -1;
	}

	// 2. Create and configure DSRC archive records writer with comp. settings:
	// lossless mode (by default) with 'fast' performance setting
	// (corresponds to '-m0' switch in DSRC binary)
	DsrcCompressionSettings settings;
	settings.dnaCompressionLevel = 0;
	settings.qualityCompressionLevel = 0;
	settings.fastqBufferSizeMb = 8;

	// initialize compression routine with prepared settings
	DsrcArchiveRecordsWriter archive;
	if (!archive.StartCompress(outFilename_, settings))
	{
		fastqFile.Close();

		std::cerr << "Error!" << std::endl;
		std::cerr << archive.GetError() << std::endl;
		return -1;
	}

	// 3. When configured, start compression
	//
	int recCount = 0;
	FastqRecord rec;
	while (fastqFile.ReadNextRecord(rec))
	{
		archive.WriteNextRecord(rec);
		recCount++;
	}

	// 4. Finish compression and close FASTQ file
	//
	archive.FinishCompress();

	fastqFile.Close();

	std::cerr << "Sucess!" << std::endl;
	std::cerr << "Compressed records: " << recCount << std::endl;
	return 0;
}


int CompressFile_lossy(const char* inFilename_, const char* outFilename_)
{
	// 1. Open FASTQ file
	//
	FastqFile fastqFile;
	if (!fastqFile.Open(inFilename_))		// TODO: analyze the dataset type
	{
		std::cerr << "Error: cannot open FASTQ file: " << inFilename_ << std::endl;
		return -1;
	}

	// 2. Create and configure DSRC archive records writer with settings:
	// use 'max' ratio compression mode
	// (corresponds to '-m2' switch in DSRC binary)
	DsrcCompressionSettings settings;
	settings.dnaCompressionLevel = 3;
	settings.qualityCompressionLevel = 2;
	settings.fastqBufferSizeMb = 256;

	// set lossy mode with Illumina 8-binning qualities scheme
	settings.lossyQualityCompression = true;

	// keep only 2 first tokens in read tag field
	settings.tagPreserveMask = FieldMask().AddField(1).AddField(2).GetMask();

	// initialize compression routine with prepared settings
	DsrcArchiveRecordsWriter archive;
	if (!archive.StartCompress(outFilename_, settings))
	{
		fastqFile.Close();

		std::cerr << "Error!" << std::endl;
		std::cerr << archive.GetError() << std::endl;
		return -1;
	}

	// 3. When configured, start compression
	//
	int recCount = 0;
	FastqRecord rec;
	while (fastqFile.ReadNextRecord(rec))
	{
		archive.WriteNextRecord(rec);
		recCount++;
	}

	// 4. Finish compression and close FASTQ file
	//
	archive.FinishCompress();

	fastqFile.Close();

	std::cerr << "Sucess!" << std::endl;
	std::cerr << "Compressed records: " << recCount << std::endl;
	return 0;
}


int DecompressFile(const char* inFilename_, const char* outFilename_)
{
	// 1. Open DSRC archive
	//
	DsrcArchiveRecordsReader archive;
	if (!archive.StartDecompress(inFilename_))
	{
		std::cerr << "Error!" << std::endl;
		std::cerr << archive.GetError() << std::endl;
		return -1;
	}

	// 2. Create output FASTQ file
	//
	FastqFile fastqFile;
	if (!fastqFile.Create(outFilename_))
	{
		archive.FinishDecompress();

		std::cerr << "Error: cannot open FASTQ file: " << inFilename_ << std::endl;
		return -1;
	}

	// 3. Start decompression
	//
	long int recCount = 0;
	FastqRecord rec;
	while (archive.ReadNextRecord(rec))
	{
		fastqFile.WriteNextRecord(rec);
		recCount++;
	}

	// 4. Finish decompression and close FASTQ file
	//
	fastqFile.Close();

	archive.FinishDecompress();

	std::cerr << "Success!" << std::endl;
	std::cerr << "Decompressed records: " << recCount << std::endl;
	return 0;
}
